provenance_lib/parse.py

```diff
@@ -0,0 +1,362 @@
+            # TODO: Should we be reading these zipfiles once here,
+            # and then passing them to the constructors below?
+            if fp.name == 'metadata.yaml':
+                self._result_md = _ResultMetadata(zf, str(fp))
+            elif fp.name == 'action.yaml':
+                self._action = _Action(zf, str(fp))
+            elif fp.name == 'citations.bib':
+                self._citations = _Citations(zf, str(fp))
+
+    def __repr__(self):
+        return f'ProvNode({self.uuid}, {self.sem_type}, fmt={self.format})'
+
+    def __str__(self):
+        return f'ProvNode({self.uuid})'
+
+    def __eq__(self, other):
+        # TODO: Should this offer more robust validation?
+        return self.uuid == other.uuid
+
+    # TODO: Should this live in ProvDAG?
+    def traverse_uuids(self):
+        """ depth-first traversal of this ProvNode's ancestors """
+        local_parents = dict()
+        if not self.parents:
+            local_parents = {self.uuid: None}
+        else:
+            sub_dag = dict()
+            for parent in self.parents:
+                sub_dag.update(parent.traverse_uuids())
+            local_parents[self.uuid] = sub_dag
+        return local_parents
+
+
+class Archive:
+    """
+    Lightly-processed contents of a single QIIME 2 Archive, an Archive
+    contains a non-hierarchical pool of unique ProvNodes
+    """
+
+    # TODO: Can we assume all Q2 archives contain only unique Results?
+    # If not, Archive should filter duplicates.
+
+    # TODO: add property for archive version?
+    # TODO: UUID class with basic validation?
+    # NOTE: static type declarations like this break testing unless the class
+    # is forward declared (ProvNode must be above Archive or NameError).
+    _number_of_results: int
+    _archive_contents: Dict[str, ProvNode]
+    _archive_md: _ResultMetadata
+
+    @property
+    def root_uuid(self):
+        return self._archive_md.uuid
+
+    @property
+    def archive_type(self):
+        return self._archive_md.type
+
+    @property
+    def archive_format(self):
+        return self._archive_md.format
+
+    def get_result(self, uuid):
+        return self._archive_contents[uuid]
+
+    def __str__(self):
+        return f"Archive(Root: {self.root_uuid})"
+
+    def __repr__(self):
+        r_str = (f"Archive(Root: {self.root_uuid}, Semantic Type: "
+                 f"{self.archive_type}, Format: {self.archive_format})\n"
+                 "Contains Results:\n")
+        for uuid in self._archive_contents.keys():
+            r_str += str(uuid)
+            r_str += "\n"
+        return r_str
+
+    def __init__(self, archive_fp: str):
+        self._archive_md: None
+        self._archive_contents: None
+        self._number_of_results = 0
+
+        # Get archive metadata including root uuid
+        with zipfile.ZipFile(archive_fp) as zf:
+            all_filenames = zf.namelist()
+            root_metadata_fps = filter(self._is_root_metadata_file,
```

Comment:
That is _way_ less gross. :laughing:

Reply:
... but isn't as straightforward as it looks to implement. 

Unless I'm overlooking something, I'd need to extract the zip archive in order to `glob` it, which is probably a bad idea. Unless there's a workaround that lets us apply `glob` to a zip archive, it's probably safer resource-wise to filter the results of `zipfile.namelist()` in one way or another.

Glob uses `fnmatch` under the hood, which offers a `filter` method that applies a nerfed regex to an iterable. This also isn't super useful; because `fnmatch` treats `*` as 0 or more of any character and doesn't seem to treat `\` as a "special" filepath separator, `fnmatch.filter(all_filenames, '*/metadata.yaml')` returns _all_ filepaths which contain `/metadata.yaml` and not just those in the top-level directory.

I could do this with `re.match`, but I'm not sure we get more value out of an explicit regex than we would out of a predicate function. 

In 8220e5e82b64b0f294631765fefbd113b276ee25, I have:

- extracted `_get_root_metadata` into its own function and cleaned up `__init__`
- simplified the `_is_root_metadata_file` predicate
- cast the filter object to a list, so we can use length checks instead of `next()` for more readable error checking

Reply:
12a2cdd6df382b2fc43b8b85b263d755efd6f155 hardcodes the root metadata filepath rather than filtering for it